### ISSUE TITLE
SEO Issues

### ISSUE THREAD

**Issue by meangczac** [PHPFusion/PHPFusion#468]
I will set a new milestone development of maximum 1 month. 

The issue currently is to 'stabilize' routing and permalink. So far, 1 issue for URL is known issue for permalink  and 1 issue is routing. Router could not fetch hardcoded CSS Files. 

SEF URL comprises of 2 parts working together.
1. Permalink - At header.php and footer.php, permalink class declare and translate rules using preg_match, preg_replace. 
2. Routing - at index.php the router must be able to use and fetch the correct file.

Both 1&2 are coded in OOP. Permalink Class and Routing Class.

OOP solution is needed because it can save parse and transfer variables among functions that need to have in order to get the whole scenario working. i.e. Using class will enable us to get shorter and more specific functions, and are generally easier to debug. 

---
# New Milestone for SEO

New Admin Panel possibly for Permalink.

Non conflicting routing using ID - like has been with AltoRouter

Major change to SEO rules - segmentize of Preg rules. (Type 1 - Number only, Type 2 - A-Z without symbol, etc and abbrieviate it as per alto router example) - like has been with AltoRouter

More as needed. 

To derive 2 most important things
1. Preparation of documentation of how to do things
2. Redo Precision in the debugger. 
   
   <ul>
   <li> What rules declared and will be used per refresh (not database but on actual page)</li>
   <li> What matched what rule </li>
   <li> What declaration has no result</li>
   </ul>

---

Tools needed:
http://www.phpliveregex.com/
http://regexr.com/
https://regex101.com/

---
# Time Schedule

Week 1 : New Admin Panel & New Permalink.
Week 2 : Routing paths development - including subdomain 
Week 3 : Permalink & Routing must be testable
Week 4 : Push and extend 1 week Beta 8 and think of a new name for this driver (Final beta)

The old set of permalink, rewrites index will not be touched. (as-is) and I will create all files new for this.

I'll start learn and train on regex this week.

**Comment by meangczac** [comment]
The intended job for week 1 was completed last night. Currently, the permalink build is not stable enough to sustain for further mods such as merging Blog category and Blog into a single permalink driver file.  I will only look into this after I double check on all existing rules. 

The new Test Build 2.0 is my preferred method to read and understand what the function does for the permalink, I rearranged 60% of the original codes and added some crude hacks into it.
The original build is 1460 lines, and the new build is around 1060 (30% lines reduction)

I see a speed change as well maybe because the calculation replacements were made in a loop earlier. This version, the replacements were condensed and start only after we prepared in full the only thing we care about - 
1. Search Regex
2. Search Replacement
3. Failed Rules

They are the new $regex_statements array. 

On a side note, I rearranged the documented comments from a massive screw up during test build 1.0x - new PHPStorm require /*\* as recognized PHPDocs comment header, not /\*

**Comment by meangczac** [comment]
The main hiccup we HAD is the failed statements, which we do not know why. 
Now, we can detect what regex statements failed. That is what we needed Day 1.  

The next step now is prepare a new page for SEO testing. I need to see what work and what not. 
a new folder will be created. /testPermalink/all_possible_blog_links.php  for test cases. I will manually hard code the links in, and when we will use this page, we will see whether it works or not. This is the best method to seperate all other issues on the actual program itself. I'll create a new issue on this.

When we have the test pages, we will reshuffle and test the rules ONE driver at a time. We find out why, and we fix it. One at a time. I don't want to go crazy.

**Comment by meangczac** [comment]
Build 2.04 - https://github.com/php-fusion/PHP-Fusion/commit/6b80a1f8ae90bf3468e10b4828d0ba80c36ab1c8

**Comment by meangczac** [comment]
2.05 & 2.06 https://github.com/php-fusion/PHP-Fusion/commit/6bf30d4dee7931a8eb895566b155540e390bf47b

Still need fixing to codes

**Comment by meangczac** [comment]
New build https://github.com/php-fusion/PHP-Fusion/commit/d231100aaf2e9944f4d168fa94249b7d51918708 3.1 now, all FUSION_SELF and Defender works without triggering Redirect_301

**Comment by meangczac** [comment]
Permalink Cycle Explained:

When Page Load, Permalink scans all .php links and translate them into SEF in its buffer. Then in one go after caching all success and failed patterns, only then it will replace HTML output. This was the major change from the old driver to the new one. I find the older is slower, tedious, and harder to debug as we cannot find source of the failed patterns via var dump..

If you use a browser .php url, Permalink will detect against its rule, by matching rule against your browser URL and redirect once found. If rule not found, the page remains in .php form. So you need to turn on and install Permalink Drivers Rule. To not waste system resource I redirect scan to runs concurrently during translating /caching SEF buffer itself - 

New driver Permalink.

```
while ( caching rules ) {
if browser is .php, find the sef url from all cached rules, if found, exit while and redirect.
}
// if redirect happens, all below will not execute - this saves lots of resources in actual replacement.
if cache rules is not empty {

while (cache_rules ) {
match original html, and replace links with sef url
}

}
```

Old Driver Permalink

```
while (caching rules) {
cache populate and compare and replace original html with sef url
}
if browser is .php find the sef url from all cached rules, and if found, redirect the page.
```

Now about router. Router is active, <strong>only</strong> whenever browser is in SEF URL format, the router will start finding the file as base include file. If filepath is not found, excepted for index.php file itself, which will auto include as opening page, otherwise, in index.php it will redirect to 404.

So if your browser is using non sef url like  /blog.php - Router dormant. It will not be found or used in any circumstances. Only Permalink is active.

What if a link is not translated? Where to look for any screw ups if  - 
a) SEF URL is used by the browser? i.e. active url is pretty url. The Router.
b) Non SEF URL is used by the browser ? The Permalink.

Will the Router redirect if non SEF URL is active? No. It's not active/sleeping.
Will the Router redirect if SEF URL is active? Yes. But Permalink is more likely to route you because it has a faster loop.

What happens when Permalink routes 301 slower than the router? 
- Forms will stop working. Post data cache will be wiped before php got executed. If Permalink is faster, it will route you to a valid $_GET request uri and perform script execution in order.
- Redirect will most likely append ../../ (this bug is fixed in router to ensure it is trimmed in Build 2+)

Why need 2 route 301? 
- Because you have non SEF Url pages and links, and SEF Url pages and links working together.

**Comment by meangczac** [comment]
Build 4.0 - https://github.com/php-fusion/PHP-Fusion/commit/603543c06db6f37b2c75217715bacd7e423b4713

Reduction of repeating codes and easier debugging process.

**Comment by meangczac** [comment]
Build 5.0 is now on, increment in progress still, but should not be far from going beta ranks.

**Comment by meangczac** [comment]
non sef url Routing is seperated.
